balancer/endpointsharding/endpointsharding.go

```diff
@@ -123,6 +124,7 @@ func (es *endpointSharding) UpdateClientConnState(state balancer.ClientConnState
 			bal.Balancer = gracefulswitch.NewBalancer(bal, es.bOpts)
 		}
 		newChildren.Set(endpoint, bal)
+		es.childMu.Lock()
```

Comment:
I think we can safely take this once at the stop and defer the unlock.  The only things that happen concurrently are:

This & Close (guaranteed to be non-concurrent) & the ExitIdle goroutine, which can be concurrent but it should be safe for that to block all of UpdateCCS.

Reply:
Oh, same thing can apply to `Close`, too, I'm pretty sure.

Reply:
ResolverError can also race with ExitIdle (all three balancer.Balancer operations downward can race). Grabbed mutex for the whole operation and added close guard.

Reply:
And I do defer an updateState() call which will now update upward with mutex held but since you mentioned can assume can't call back inline I'll go ahead and add it.